### DISCUSSION TITLE
Question about treadmill powerzones

### DISCUSSION THREAD

**Discussion by lorenzogavanna** [cagnulein/qdomyos-zwift#2873] [General]
Ciao Roberto, 
`            if (atts.hasAttribute(QStringLiteral("powerzone"))) {`
I little question how these powerzones are meant for treadmill? They are for sure realted to the "Critical Power Run value" setting, but I can't get how the standard 7 zones are distribuited in the calculation and which are the value to put in the xml tag. Thank you!! 
file [trainprogram.cpp](https://github.com/cagnulein/qdomyos-zwift/blob/301429182d4089cca108e236840bfeaa25c6d8d0/src/trainprogram.cpp#L1557)
P.S. your xml seems to be very powerful and with much more tags than the ones you shared online!!!! Trying to figure out everything to give a "manual" for everybody

**Comment by cagnulein** [comment]
hi @lorenzogavanna 

all is about this line

`row.power = atts.value(QStringLiteral("powerzone")).toDouble() * settings.value(QZSettings::ftp_run, QZSettings::default_ftp_run).toDouble();`

so it's scaled to the value from the zwo file in powerzone * your critucal power run value

**Comment by lorenzogavanna** [comment]
Ok, thank you! And for the xml, is this the same? Because if I put a value
like <row duration="00:01:00" powerzone="4"/> this should be around 8km/h
with my settings (Power run, and weight) but qz goes at 15km/h. Let me know
if you need more tests.


---
Lorenzo Gavanna
www.lorenzogavanna.com
***@***.***
+39 347 6788352


Il giorno mer 4 dic 2024 alle ore 09:39 Roberto Viola <
***@***.***> ha scritto:

**Reply by cagnulein** [reply]
yes it should be the same. which ftp do you have?

**Reply by lorenzogavanna** [reply]
200. let me share a log with you.
[debug-Wed_Dec_4_09_57_04_2024.log](https://github.com/user-attachments/files/18005305/debug-Wed_Dec_4_09_57_04_2024.log)
[11d.xml.txt](https://github.com/user-attachments/files/18005307/11d.xml.txt)

**Reply by cagnulein** [reply]
i guess the issue is that you have to scale it by /10. I mean 4 * 200 = 800 W so that's the issue. I guess you should use 0.4 or a similar value.

I used the same approach that is for rides

**Reply by lorenzogavanna** [reply]
I tried that as well, it was one of my guess before writing you. If I put 0.4 (that could make sense) the speed is 3,5 km/h, that is far away from 8 kmh (that is the one for zone 4).
Now I think I got it. The right speed is reached when is about 1. So it should be that way:
since zwift zones are like this
```
Zone 1 (Grey, Recovery): below 60%
Zone 2 (Blue, Endurance): 60-75%
Zone 3 (Green, Tempo): 76-89%
Zone 4 (Yellow, Threshold): 90-104%
Zone 5 (Orange, VO2 Max): 105-118%
Zone 6 (Red, Anaerobic): above 118%
```

those are the values you need to put in the xml
```
Zone 1 (Grey, Recovery): below 0.6
Zone 2 (Blue, Endurance): 0.6-0.75
Zone 3 (Green, Tempo): 0.76-0.89
Zone 4 (Yellow, Threshold): 0.90-1.04
Zone 5 (Orange, VO2 Max): 1.05-1.18
Zone 6 (Red, Anaerobic): above 1.18
```
do you think it makes sense?

**Reply by cagnulein** [reply]
yes because they are more or less the values that i wrote here for the bike

```
if (ftpSetting > 0) {
            ftpPerc = (watts / ftpSetting) * 100.0;
            if (bluetoothManager->device()->deviceType() == bluetoothdevice::BIKE) {
                requestedPerc =
                    (((bike *)bluetoothManager->device())->lastRequestedPower().value() / ftpSetting) * 100.0;
            } else if (bluetoothManager->device()->deviceType() == bluetoothdevice::ROWING) {
                requestedPerc =
                    (((rower *)bluetoothManager->device())->lastRequestedPower().value() / ftpSetting) * 100.0;
            }
        }
        if (ftpPerc < 56) {
            ftpMinW = QString::number(0, 'f', 0);
            ftpMaxW = QString::number(ftpSetting * 0.55, 'f', 0);
            ftpZone = 1;
            ftpZone += (ftpPerc / 56);
            if (ftpZone >= 1.95) { // double precision could cause unwanted approximation
                ftpZone = 1.9;
            }
            ftp->setValueFontColor(QStringLiteral("white"));
            watt->setValueFontColor(QStringLiteral("white"));
        } else if (ftpPerc < 76) {

            ftpMinW = QString::number((ftpSetting * 0.55) + 1, 'f', 0);
            ftpMaxW = QString::number(ftpSetting * 0.75, 'f', 0);
            ftpZone = 2;
            ftpZone += ((ftpPerc - 56) / 20);
            if (ftpZone >= 2.95) { // double precision could cause unwanted approximation
                ftpZone = 2.9;
            }
            ftp->setValueFontColor(QStringLiteral("limegreen"));
            watt->setValueFontColor(QStringLiteral("limegreen"));
        } else if (ftpPerc < 91) {

            ftpMinW = QString::number((ftpSetting * 0.75) + 1, 'f', 0);
            ftpMaxW = QString::number(ftpSetting * 0.90, 'f', 0);
            ftpZone = 3;
            ftpZone += ((ftpPerc - 76) / 15);
            if (ftpZone >= 3.95) { // double precision could cause unwanted approximation
                ftpZone = 3.9;
            }
            ftp->setValueFontColor(QStringLiteral("gold"));
            watt->setValueFontColor(QStringLiteral("gold"));
        } else if (ftpPerc < 106) {

            ftpMinW = QString::number((ftpSetting * 0.90) + 1, 'f', 0);
            ftpMaxW = QString::number(ftpSetting * 1.05, 'f', 0);
            ftpZone = 4;
            ftpZone += ((ftpPerc - 91) / 15);
            if (ftpZone >= 4.95) { // double precision could cause unwanted approximation
                ftpZone = 4.9;
            }
            ftp->setValueFontColor(QStringLiteral("orange"));
            watt->setValueFontColor(QStringLiteral("orange"));
        } else if (ftpPerc < 121) {

            ftpMinW = QString::number((ftpSetting * 1.05) + 1, 'f', 0);
            ftpMaxW = QString::number(ftpSetting * 1.20, 'f', 0);
            ftpZone = 5;
            ftpZone += ((ftpPerc - 106) / 15);
            if (ftpZone >= 5.95) { // double precision could cause unwanted approximation
                ftpZone = 5.9;
            }
            ftp->setValueFontColor(QStringLiteral("darkorange"));
            watt->setValueFontColor(QStringLiteral("darkorange"));
        } else if (ftpPerc < 151) {

            ftpMinW = QString::number((ftpSetting * 1.20) + 1, 'f', 0);
            ftpMaxW = QString::number(ftpSetting * 1.50, 'f', 0);
            ftpZone = 6;
            ftpZone += ((ftpPerc - 121) / 30);
            if (ftpZone >= 6.95) { // double precision could cause unwanted approximation
                ftpZone = 6.9;
            }
            ftp->setValueFontColor(QStringLiteral("orangered"));
            watt->setValueFontColor(QStringLiteral("orangered"));
        } else {

            ftpMinW = QString::number((ftpSetting * 1.50) + 1, 'f', 0);
            ftpMaxW = QStringLiteral("∞");
            ftpZone = 7;

            ftp->setValueFontColor(QStringLiteral("red"));
            watt->setValueFontColor(QStringLiteral("red"));
        }
```


do you see the 0.55 and so on?

**Reply by lorenzogavanna** [reply]
yes, thank you! Perfect so that's why it was not working. The powerzone tag is not the actual zone but the number % in the zone (based on zwift zones reference). Fantastic! I think that, even you, don't know how potential there is in your xml ehehe. You can have ramps, (based on zones or speed), distances and time rows combined in the same workout. Actually is much better than the zwo! 
Probably I will implement something starting from your workout creator html and I will be back to you in weeks. Thank you Roberto!

**Comment by lorenzogavanna** [comment]
Do you know why the last section of this workout had the duration wrong? It last 3 mins and 45 seconds instead of only 3. Let me know if you need more test or if you want that I write a new post on Github
[workout.xml.txt](https://github.com/user-attachments/files/18007800/workout.xml.txt)
[debug-Wed_Dec_4_12_14_41_2024.log](https://github.com/user-attachments/files/18007801/debug-Wed_Dec_4_12_14_41_2024.log)

**Reply by cagnulein** [reply]
ops fixed https://github.com/cagnulein/qdomyos-zwift/actions/runs/12159803163 @lorenzogavanna thanks!

**Comment by lorenzogavanna** [comment]
thank you!

**Comment by lorenzogavanna** [comment]
I loaded it again. And now the problem was at the beginning it was 00:01:51 (in the file is 00:02:00). Didn't try it until the end. Let me know if you need deep test.

**Reply by cagnulein** [reply]
in your file the first one was of 1:30? am i missing something?

**Comment by lorenzogavanna** [comment]
No my bad you're right, I said 2 and I was wrong. But qz showed me 00:01:51
---
Lorenzo Gavanna
www.lorenzogavanna.com
***@***.***
+39 347 6788352

Il mer 4 dic 2024, 14:36 Roberto Viola ***@***.***> ha
scritto:

**Reply by cagnulein** [reply]
ok i will check it

**Reply by lorenzogavanna** [reply]
Little add: it happens only with powerzonefrom powerzoneto rows. If you put speedfrom and speedto tags it loads and calculate the duration correctly!

**Reply by cagnulein** [reply]
sorry i still can't understand where do you see the 1:51?

i checked the program loaded with the new version and it is

```
mer dic 4 15:03:36 2024 1733321016460 Debug: ..\src\trainprogram.cpp static QList<trainrow> trainprogram::loadXML(const QString&, bluetoothdevice::BLUETOOTH_TYPE) TrainRow "duration = 00:00:22 distance = -1 speed = -1 lower_speed = -1 average_speed = -1 upper_speed = -1 fanspeed = -1 inclination = -200 lower_inclination = -200 average_inclination = -200 upper_inclination = -200 resistance = -1 lower_resistance = -1 average_resistance = -1 upper_resistance = -1 requested_peloton_resistance = -1 lower_requested_peloton_resistance = -1 average_requested_peloton_resistance = -1 upper_requested_peloton_resistance = -1 pace_intensity = -1 cadence = -1 lower_cadence = -1 average_cadence = -1 upper_cadence = -1 forcespeed = 1 loopTimeHR = 10 zoneHR = -1 HRmin = -1 HRmax = -1 maxSpeed = -1 minSpeed = -1 maxResistance = -1 power = 110 mets = -1 latitude = nan longitude = nan altitude = nan azimuth = nan rampElapsed = 00:00:00 rampDuration = 00:01:30"
mer dic 4 15:03:36 2024 1733321016461 Debug: ..\src\trainprogram.cpp static QList<trainrow> trainprogram::loadXML(const QString&, bluetoothdevice::BLUETOOTH_TYPE) TrainRow "duration = 00:00:22 distance = -1 speed = -1 lower_speed = -1 average_speed = -1 upper_speed = -1 fanspeed = -1 inclination = -200 lower_inclination = -200 average_inclination = -200 upper_inclination = -200 resistance = -1 lower_resistance = -1 average_resistance = -1 upper_resistance = -1 requested_peloton_resistance = -1 lower_requested_peloton_resistance = -1 average_requested_peloton_resistance = -1 upper_requested_peloton_resistance = -1 pace_intensity = -1 cadence = -1 lower_cadence = -1 average_cadence = -1 upper_cadence = -1 forcespeed = 1 loopTimeHR = 10 zoneHR = -1 HRmin = -1 HRmax = -1 maxSpeed = -1 minSpeed = -1 maxResistance = -1 power = 112 mets = -1 latitude = nan longitude = nan altitude = nan azimuth = nan rampElapsed = 00:00:22 rampDuration = 00:01:08"
mer dic 4 15:03:36 2024 1733321016462 Debug: ..\src\trainprogram.cpp static QList<trainrow> trainprogram::loadXML(const QString&, bluetoothdevice::BLUETOOTH_TYPE) TrainRow "duration = 00:00:23 distance = -1 speed = -1 lower_speed = -1 average_speed = -1 upper_speed = -1 fanspeed = -1 inclination = -200 lower_inclination = -200 average_inclination = -200 upper_inclination = -200 resistance = -1 lower_resistance = -1 average_resistance = -1 upper_resistance = -1 requested_peloton_resistance = -1 lower_requested_peloton_resistance = -1 average_requested_peloton_resistance = -1 upper_requested_peloton_resistance = -1 pace_intensity = -1 cadence = -1 lower_cadence = -1 average_cadence = -1 upper_cadence = -1 forcespeed = 1 loopTimeHR = 10 zoneHR = -1 HRmin = -1 HRmax = -1 maxSpeed = -1 minSpeed = -1 maxResistance = -1 power = 114 mets = -1 latitude = nan longitude = nan altitude = nan azimuth = nan rampElapsed = 00:00:45 rampDuration = 00:00:45"
mer dic 4 15:03:36 2024 1733321016464 Debug: ..\src\trainprogram.cpp static QList<trainrow> trainprogram::loadXML(const QString&, bluetoothdevice::BLUETOOTH_TYPE) TrainRow "duration = 00:00:22 distance = -1 speed = -1 lower_speed = -1 average_speed = -1 upper_speed = -1 fanspeed = -1 inclination = -200 lower_inclination = -200 average_inclination = -200 upper_inclination = -200 resistance = -1 lower_resistance = -1 average_resistance = -1 upper_resistance = -1 requested_peloton_resistance = -1 lower_requested_peloton_resistance = -1 average_requested_peloton_resistance = -1 upper_requested_peloton_resistance = -1 pace_intensity = -1 cadence = -1 lower_cadence = -1 average_cadence = -1 upper_cadence = -1 forcespeed = 1 loopTimeHR = 10 zoneHR = -1 HRmin = -1 HRmax = -1 maxSpeed = -1 minSpeed = -1 maxResistance = -1 power = 116 mets = -1 latitude = nan longitude = nan altitude = nan azimuth = nan rampElapsed = 00:01:07 rampDuration = 00:00:23"
```

so are 3 rows of 22 seconds + 1 of 23 seconds. are you seeing that number on the qz tile?

**Reply by lorenzogavanna** [reply]
yes in the tile,

**Reply by cagnulein** [reply]
hah saw now also on the tile. let me check

**Reply by cagnulein** [reply]
it was only a UI visualization error, it should be fixed now https://github.com/cagnulein/qdomyos-zwift/actions/runs/12162138531
unfortunately i don't have a lot of time for testing it

**Comment by lorenzogavanna** [comment]
Thank you! later today i will do a workout will be back to you!
---
Lorenzo Gavanna
www.lorenzogavanna.com
***@***.***
+39 347 6788352

Il mer 4 dic 2024, 15:46 Roberto Viola ***@***.***> ha
scritto:

**Comment by lorenzogavanna** [comment]
Hi Roberto,
everything seems to work properly (I double tested it). The only thing is
that the duration of the two ramps (the first and last row of the workout)
were shown off by 1 second in the tile. 00:01:30--> 00:01:29 and
00:03:00-->00:02:59 in the tile. Don't think you need to recompile
everything just for that. Everything is ok! Thank you!


---
Lorenzo Gavanna
www.lorenzogavanna.com
***@***.***
+39 347 6788352


Il giorno mer 4 dic 2024 alle ore 15:47 Lorenzo Gavanna <
***@***.***> ha scritto:

**Comment by cagnulein** [comment]
yes i saw that too, i have to verify where that difference comes from
thanks again for the test

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 5 dic 2024 alle ore 09:06 lorenzogavanna <
***@***.***> ha scritto:
